File: rust/examples/linux/pressure/bmp085/demo/src/main.rs
use linux_embedded_hal::I2cdev;
use periph::chips::pressure::{Bmp085Full, OSS_ULP};

fn main() {
    let i2c_bus: u8 = std::env::var("I2C_BUS").ok().and_then(|v| v.parse().ok()).unwrap_or(1);

    let dev = I2cdev::new(format!("/dev/i2c-{}", i2c_bus)).expect("open i2c bus");
    let mut bmp = Bmp085Full::new(dev, 0x77, OSS_ULP).expect("init BMP085"); // Create BMP085 driver, (connection, oss=0 ULP)

    let t0 = bmp.temperature().expect("read temperature");                   // Read temperature, () → f32 C
    let p0 = bmp.pressure().expect("read pressure");                        // Read pressure, () → f32 Pa
    let alt_ref = bmp.altitude(101325.0).expect("compute altitude");       // Compute altitude, (sea_level_pa=101325.0) → f32 m
    println!("Reference: {:.1} C, {:.1} Pa, alt={:.1} m", t0, p0, alt_ref);
    let mut prev_alt = 0.0f32;

    for n in 0..60 {
        let t = bmp.temperature().expect("read temperature");               // Read temperature, () → f32 C
        let p = bmp.pressure().expect("read pressure");                     // Read pressure, () → f32 Pa
        let a = bmp.altitude(101325.0).expect("compute altitude");         // Compute altitude, (sea_level_pa=101325.0) → f32 m
        let da = (a - prev_alt) * 100.0;

        if n > 0 {
            println!("{}s: {:.1} C, {:.1} Pa, alt={:.1} m (delta={:+.0} cm)", n, t, p, a, da);
        } else {
            println!("{}s: {:.1} C, {:.1} Pa, alt={:.1} m", n, t, p, a);
        }
        prev_alt = a;
        std::thread::sleep(std::time::Duration::from_secs(1));
    }
}

File: cpp/examples/linux/pressure/BMP085/demo/main.cpp
#ifndef TEST_I2C_BUS
#define TEST_I2C_BUS 1
#endif
#ifndef TEST_ADDR
#define TEST_ADDR 0x77
#endif

#include <cstdio>
#include <chrono>
#include <thread>
#include "I2CConnectionLinux.h"
#include "BMP085.h"

int main() {
    I2CConnectionLinux connection(TEST_I2C_BUS, TEST_ADDR);
    BMP085Full bmp(connection, BMP085Full::OSS_ULP);        // Create BMP085 driver, (connection, oss=0 ULP)

    float t0 = bmp.temperature();                           // Read temperature, () → float C
    float p0 = bmp.pressure();                             // Read pressure, () → float Pa
    float alt_ref = bmp.altitude();                        // Compute altitude, (sea_level_pa=101325.0) → float m
    printf("Reference: %.1f C, %.1f Pa, alt=%.1f m\n", t0, p0, alt_ref);

    float prev_alt = 0.0f;
    for (int n = 0; n < 60; n++) {
        float t = bmp.temperature();                        // Read temperature, () → float C
        float p = bmp.pressure();                          // Read pressure, () → float Pa
        float a = bmp.altitude();                          // Compute altitude, (sea_level_pa=101325.0) → float m
        float da = (a - prev_alt) * 100.0f;

        if (n > 0) {
            printf("%ds: %.1f C, %.1f Pa, alt=%.1f m (delta=%+.0f cm)\n", n, t, p, a, da);
        } else {
            printf("%ds: %.1f C, %.1f Pa, alt=%.1f m\n", n, t, p, a);
        }
        prev_alt = a;
        std::this_thread::sleep_for(std::chrono::seconds(1));
    }

    printf("===DONE: 0 passed, 0 failed===\n");
    return 0;
}
